File: LAB3/CODE/celebrityproblem1.cpp
#include <iostream>
#include <stack>
using namespace std;

int celebrityproblemfunction(int arr[2][2], int n) {
    stack<int> s;

    for(int i = 0; i < n; i++)
        s.push(i);

    while(s.size() > 1) {
        int a = s.top();
         s.pop();
        int b = s.top(); 
        s.pop();

        if(arr[a][b] == 1) 
            s.push(b);
        else               
            s.push(a);
    }

    int candidate = s.top();

    for(int i = 0; i < n; i++) {
        if(i != candidate && (arr[candidate][i] == 1 || arr[i][candidate] == 0))
            return -1; 
    }

    return candidate;
}

int main() {
    const int n = 2;
    int arr[n][n];

    cout << "Enter the 2x2 matrix:\n";
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            cin >> arr[i][j];

    int celeb = celebrityproblemfunction(arr, n);

    if(celeb == -1)
        cout << "No celebrity found\n";
    else
        cout << "Celebrity is person " << celeb << endl;

    return 0;
}

File: LAB3/CODE/celebrityproblem1.java
import java.util.Scanner;
import java.util.Stack;

public class celebrityproblem1{

    public static int celebrityProblemFunction(int[][] arr, int n) {
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            stack.push(i);
        }
        while (stack.size() > 1) {
            int a = stack.pop();
            int b = stack.pop();

            if (arr[a][b] == 1) {
                stack.push(b);
            } else {
                stack.push(a);
            }
        }

        int candidate = stack.pop();
        for (int i = 0; i < n; i++) {
            if (i != candidate && (arr[candidate][i] == 1 || arr[i][candidate] == 0)) {
                return -1; 
            }
        }

        return candidate;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int n = 2;
        int[][] arr = new int[n][n];

        System.out.println("Enter the 2x2 matrix:");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                arr[i][j] = sc.nextInt();
            }
        }

        int celeb = celebrityProblemFunction(arr, n);

        if (celeb == -1) {
            System.out.println("No celebrity found");
        } else {
            System.out.println("Celebrity is person " + celeb);
        }

        sc.close();
    }
}
